File: core/src/ru/myitschool/dcrawler/effects/BattlecryEffect.java
package ru.myitschool.dcrawler.effects;

import com.badlogic.gdx.graphics.Texture;
import ru.myitschool.dcrawler.entities.Entity;
import ru.myitschool.dcrawler.math.MathAction;
import ru.myitschool.dcrawler.math.SumAction;

/**
 * Created by Voyager on 19.11.2017.
 */
public class BattlecryEffect extends Effect {

    private String name = "Battle cry";
    private String description = "";
    private Texture icon = new Texture("battle_cry.png");
    private boolean used = false;
    private boolean positive = true;
    private MathAction damage;
    private int accuracyBonus;
    private int turns;

    public BattlecryEffect(Entity entity, int accuracyBonus, MathAction damage, int turns) {
        super(entity);
        this.turns = turns;
        this.damage = damage;
        this.accuracyBonus = accuracyBonus;
        entity.setAccuracyBonus(entity.getAccuracyBonus() + accuracyBonus);
    }

    @Override
    public MathAction attackBonus(MathAction action) {
        action = super.attackBonus(action);
        MathAction resultAction = new SumAction(action, damage);
        return resultAction;
    }

    @Override
    public Texture getIcon() {
        return icon;
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public String getDescription() {
        return "Increases accuracy for " + accuracyBonus + " and damage for " + damage.getDescription() + " for " + turns + " turns.";
    }

    @Override
    public boolean isSkillUse() {
        return false;
    }

    @Override
    public boolean isPositive() {
        return true;
    }

    @Override
    public boolean isStackable() {
        return false;
    }

    @Override
    public int getStackSize() {
        return 0;
    }

    @Override
    public boolean isExpiring() {
        return true;
    }

    @Override
    public int getExpireTurns() {
        return turns;
    }

    @Override
    public String getType() {
        return "main.dcrawler.effect.battlecry";
    }

    @Override
    public void startTurn() {
        super.startTurn();
        if (getExpireTurns() == 0){
            getEntity().setAccuracyBonus(getEntity().getAccuracyBonus() - accuracyBonus);
        }
    }
}

File: core/src/com/wolg_vlad/dcrawler/effects/BattlecryEffect.kt
package com.wolg_vlad.dcrawler.effects

import com.badlogic.gdx.graphics.Texture
import com.wolg_vlad.dcrawler.entities.Entity
import com.wolg_vlad.dcrawler.math.MathAction
import com.wolg_vlad.dcrawler.math.SumAction

/**
 * Created by Voyager on 19.11.2017.
 */
class BattlecryEffect(entity: Entity?, private val accuracyBonus: Int, private val damage: MathAction, private val turns: Int) : Effect(entity) {
    override val name = "Battle cry"
    override val description = "Increases accuracy for " + accuracyBonus + " and damage for " + damage.description + " for " + turns + " turns."
    override val icon = Texture("battle_cry.png")
    private val used = false
    override val positive: Boolean = true
    override val stackable: Boolean = false
    override val skillUse: Boolean = false
    override val stackSize: Int = 0
    override val expiring: Boolean = true
    override val expireTurns: Int = turns
    override val id: String = "main.dcrawler.effect.battlecry"

    override fun attackBonus(action: MathAction): MathAction {
        var action = action
        action = super.attackBonus(action)
        return SumAction(action, damage)
    }

    override fun startTurn() {
        super.startTurn()
        if (expireTurns == 0) {
            entity!!.accuracyBonus = entity!!.accuracyBonus - accuracyBonus
        }
    }

    init {
        entity!!.accuracyBonus = entity.accuracyBonus + accuracyBonus
    }
}